File: python/2024/day02.py
from itertools import pairwise


def inspect_report(report: list[int]) -> bool:
    if report[0] < report[-1]:
        return all(a < b and b - a <= 3 for a, b in pairwise(report))
    else:
        return all(a > b and a - b <= 3 for a, b in pairwise(report))


class State:
    def __init__(self, input: str) -> None:
        self.reports = [list(map(int, line.split(" "))) for line in input.splitlines()]

    def safe_reports(self) -> int:
        return len(list(filter(inspect_report, self.reports)))

    def tolerable_reports(self) -> int:
        def f(report: list[int]) -> bool:
            return any(inspect_report(report[:i] + report[i + 1 :]) for i in range(len(report)))

        return len(list(filter(f, self.reports)))


def part1(input: str) -> int:
    return State(input).safe_reports()


def part2(input: str) -> int:
    return State(input).tolerable_reports()


TEST_INPUT = """7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9"""

PART1_TESTS = [
    (TEST_INPUT, 2),
]

PART2_TESTS = [
    (TEST_INPUT, 4),
]

File: rust/2024/src/bin/day02.rs
fn inspect_report(report: &[u8]) -> bool {
    if report[0] < report[report.len() - 1] {
        report.windows(2).all(|v| v[0] < v[1] && v[1] - v[0] <= 3)
    } else {
        report.windows(2).all(|v| v[0] > v[1] && v[0] - v[1] <= 3)
    }
}

#[derive(Clone)]
struct Solution {
    reports: Vec<Vec<u8>>,
}

impl Solver for Solution {
    fn new(input: &str) -> Anyhow<Self> {
        let mut reports = Vec::with_capacity(input.lines().count());
        let mut report = Vec::new();
        let mut bytes = input.bytes();

        while let Some(b) = bytes.next() {
            let mut num = b - b'0';

            if let Some(mut b) = bytes.next() {
                if b.is_ascii_digit() {
                    num = (num << 1) + (num << 3) + b - b'0';
                    b = bytes.next().unwrap();
                }

                report.push(num);

                if b == b'\n' {
                    reports.push(report.clone());
                    report.clear();
                }
            }
        }

        Ok(Self { reports })
    }

    fn part1(&mut self) -> Anyhow<impl fmt::Display> {
        Ok(self
            .reports
            .iter()
            .filter(|report| inspect_report(report))
            .count())
    }

    fn part2(&mut self) -> Anyhow<impl fmt::Display> {
        Ok(self
            .reports
            .par_iter()
            .filter(|report| {
                (0..report.len()).any(|i| {
                    inspect_report(&[&report[0..i], &report[i + 1..report.len()]].concat())
                })
            })
            .count())
    }
}

aoc::solution!();

#[cfg(test)]
mod test {
    use super::{Solution, Solver};

    const INPUT: &str = r"7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9
";

    #[test]
    fn test_part1() {
        let mut solution = Solution::new(INPUT).unwrap();
        let answer = solution.part1().unwrap().to_string();
        assert_eq!(answer, "2");
    }

    #[test]
    fn test_part2() {
        let mut solution = Solution::new(INPUT).unwrap();
        let answer = solution.part2().unwrap().to_string();
        assert_eq!(answer, "4");
    }
}
